var/spack/repos/builtin/packages/py-cffi/package.py

```diff
@@ -32,12 +33,22 @@ class PyCffi(PythonPackage):
     # ./spack-src/cffi/ffiplatform.py has _hack_at_distutils which imports
     # setuptools before distutils, but only on Windows. This could be made
     # unconditional to support Python 3.12
-    depends_on("python@:3.11", type=("build", "run"))
+    depends_on("python@:3.11", type=("build", "run"), when="@:1.15")
+    depends_on("python@:3.12", type=("build", "run"), when="@1.16:")
```

Comment:
I don't think this is correct, Python 3.12 is still not supported last I checked, even in 1.17: #46224

Reply:
This is incorrect, see https://github.com/cython/cython/pull/5754#issuecomment-1752102480

Reply:
I've read through both of those (thanks!). So, `cffi` /does/ support python 3.12, but you believe it doesn't in a way that is robust for spack (or technically correct) -- correct? 

Indeed, I can build and cleanly import `py-cffi ^python@3.12`. I haven't tested this completely. Are you saying that parts of it don't behave as expected in the spack prefix? In your closed PR what problems did you see?

Looking at `cffi 0.16:`, `cffi` has a specific shim to work around the python 3.12 disutils removal
https://github.com/python-cffi/cffi/blob/release-1.16/src/cffi/_shimmed_dist_utils.py
that looks to do what @haampie  suggested in https://github.com/cython/cython/pull/5754#issuecomment-1752102480

Reply:
I'm fine to remove everything from this PR except the patch application. But, imo, it would be good to also address this 3.12 issue if cffi really isn't working, or at least put a comment into it such that some other poor person doesn't waste their time on this until the dust settles.

Reply:
I agree, I would love to have 3.12 support. Last time I tried, cffi builds fine, but nothing else builds with it. You should see CI failures if I'm right.

Reply:
I modified `py-numpy`'s python upper limited to 3.13, and installed `py-cfgrib` against this `py-cffi` build. I was able to load a grib file, and this is a path that touches `cffi` (as per this PR's patch). Is this unexpected given your experience?

Reply:
Cffi is a big library, and only certain parts import distutils. If you can find a way to get CI to pass, then we're good to go.

Reply:
The CI failure seems to be "just" flux-core of which I have zero insight. It doesn't build on macos so it requires more effort to debug. I don't have the time to chase 3.12 at the moment, so I have removed the python@3.12 support and added a comment to save the next person some time.

Reply:
I don't know anything about flux-core either, maybe @grondo can comment on the cffi detection mechanism and why it might not be working?